cluster/src/main/java/org/apache/iotdb/cluster/log/catchup/CatchUpTask.java

```diff
@@ -0,0 +1,169 @@
+            peer.getNextIndex() + 1);
+      } catch (Exception e) {
+        logger.error("Unexpected error in logManager's getEntries during matchIndexCheck", e);
+      }
+    }
+
+    int index = logs.size() - 1;
+    EmptyContentLog emptyLog = new EmptyContentLog();
+    while (index >= 0) {
+      Log log = logs.get(index);
+      emptyLog.setCurrLogIndex(log.getCurrLogIndex());
+      emptyLog.setCurrLogTerm(log.getCurrLogTerm());
+      synchronized (raftMember.getTerm()) {
+        // make sure this node is still a leader
+        if (raftMember.getCharacter() != NodeCharacter.LEADER) {
+          logger.debug("Leadership is lost when doing a catch-up to {}, aborting", node);
+          break;
+        }
+        request.setTerm(raftMember.getTerm().get());
+      }
+      if (index > 0) {
+        request.setPrevLogIndex(logs.get(index - 1).getCurrLogIndex());
+        request.setPrevLogTerm(logs.get(index - 1).getCurrLogTerm());
+      } else {
+        request.setPrevLogIndex(log.getCurrLogIndex() - 1);
+        try {
+          request.setPrevLogTerm(raftMember.getLogManager().getTerm(log.getCurrLogIndex() - 1));
+        } catch (Exception e) {
+          logger.error("getTerm failed for newly append entries", e);
+        }
+      }
+
+      handler.setLog(log);
+      request.setEntry(emptyLog.serialize());
+
+      synchronized (isMatch) {
+        AsyncClient client = raftMember.connectNode(node);
+        if (client == null) {
+          break;
+        }
+        client.appendEntry(request, handler);
+        raftMember.getLastCatchUpResponseTime().put(node, System.currentTimeMillis());
+        isMatch.wait(RaftServer.connectionTimeoutInMS);
+      }
+      if (isMatch.get()) {
+        // if follower return RESPONSE.AGREE with this empty log, then start sending real logs from this emptyContentLog's index.
+        // which means the log which index is matchIndex will be send twice, but at the first time it sent an empty log.
+        logs.subList(0, index).clear();
+        if (logger.isDebugEnabled()) {
+          logger.debug("{} makes {} catch up with {} cached logs", raftMember.getName(), node,
+              logs.size());
+        }
+        return false;
+      }
+      index--;
+    }
+    try {
+      raftMember.getLogManager().takeSnapshot();
+    } catch (IOException e) {
+      logger.error("Unexpected error when taking snapshot.", e);
+    }
+    snapshot = raftMember.getLogManager().getSnapshot();
+    if (logger.isDebugEnabled()) {
+      logger
+          .debug("{}: Logs in {} are too old, catch up with snapshot", raftMember.getName(), node);
+    }
+    return true;
+  }
+
+  public void run() {
+    try {
+      if (checkMatchIndex()) {
+        SnapshotCatchUpTask task = new SnapshotCatchUpTask(logs, snapshot, node, raftMember);
+        task.run();
+      } else {
+        LogCatchUpTask task = new LogCatchUpTask(logs, node, raftMember);
+        task.run();
+      }
```

Comment:
Batched log catch-up has been introduced, better to switch to better.

Reply:
Batched log catch-up or not is only related to the configuration item, and the code here supports batched log catch-up if related configuration item has been set~